Improve status information from Shadow Link

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR introduces a new resource in the proto called `ShadowTopic` and some new RPCs to list and get `ShadowTopics` from `ShadowLink`'s.  The `ShadowTopic` will contain information about the _source_ partition's offsets and a timestamp for when the last time the offsets were updated.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
pb: Create ShadowTopic resource

Created the ShadowTopic resource which is a sub-resource of the
ShadowLink resource.  This is separated out to allow for RPCs to be
written to directly fetch this resource.  Additionally, when users are
given the ability to manually created ShadowTopics, this will be done
using this resource.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
proto: Added RPCs for ShadowTopic resource

Added RPCs to get and list ShadowTopics

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
admin/sl: Implement the List and Get RPCs

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: introduce replication types

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Add ability to get source partition information

Add ability through the CL service the ability to get the source topic
partition.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Return if error from topic report

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/model: Add source parition offsets to report

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/model: Added new messages for full shadow link report

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: shard_local report does not need to be async

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/model: Added status report type

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Add methods to get link report

Added ability to get link reports per node and shard.  Future commits
will call these methods to get the whole cluster report.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/rpc: Add new rpc service to get link report

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Add ability to fetch link status

Added methods to gather the status of all shadow topics on a link

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
admin/sl: Improve status information

Status information about shadow topics has been improved to contain
information about the source partitions offsets

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
dt/cl: Added tests to use new Shadow Topic services

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
/dt

michael-redpanda commented on tests/java/e2e-verifiers/dependency-reduced-pom.xml:
.... how did this get in here?   i'll get rid of it lol

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--dc3c04bb-064c-4700-9981-9a2034147441-->
<!--TEST RESULTS-->
<!--af0fd94a-80c4-4e96-8f0e-5528f1b95c71-->
<!--TEST RESULTS-->
<!--9491298c-3684-4623-9745-bbb24c9d830d-->
<!--TEST RESULTS-->
<!--2ea8b26f-45d7-4381-b2c1-37457fb22bb9-->
<!--TEST RESULTS-->
<!--549cfff3-fce0-4b76-91a5-6b6403d7431b-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73863">build#73863</a></summary>


| test_class                    | test_method              | test_arguments                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 | test_history                                                                                                                                                        |
|:------------------------------|:-------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkTopicFailoverTests  | test_link_failover       | {"with_failures": false}                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/73863#0199c9c5-7da4-4b61-abf5-db0de6aa572c | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkTopicFailoverTests&test_method=test_link_failover       |
| ShadowLinkTopicFailoverTests  | test_link_failover       | {"with_failures": true}                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/73863#0199c9c5-7da4-4caf-8b24-490235ac60fb | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkTopicFailoverTests&test_method=test_link_failover       |
| ShadowLinkTopicFailoverTests  | test_link_topic_failover | {"with_failures": false}                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/73863#0199c9c5-7da6-41fa-bef3-771bc9746438 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkTopicFailoverTests&test_method=test_link_topic_failover |
| ShadowLinkTopicFailoverTests  | test_link_topic_failover | {"with_failures": true}                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/73863#0199c9c5-7da7-4593-b460-e09d66e9ce6b | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkTopicFailoverTests&test_method=test_link_topic_failover |
| ShadowLinkingReplicationTests | test_replication_basic   | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73863#0199c9c5-7da9-4ae2-adbc-6005e2f1fef8 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic  |
| ShadowLinkingReplicationTests | test_replication_basic   | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/73863#0199c9c5-7daa-4900-958d-1a21aab07f11 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic  |
| ShadowLinkingReplicationTests | test_topic_delete        | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}}                             | integration | https://buildkite.com/redpanda/redpanda/builds/73863#0199c9c5-7da4-4b61-abf5-db0de6aa572c | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_topic_delete       |
| ShadowLinkingReplicationTests | test_topic_delete        | {"source_cluster_spec": {"cluster_type": "redpanda"}}                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/73863#0199c9c5-7da4-4caf-8b24-490235ac60fb | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_topic_delete       |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74086">build#74086</a></summary>


| test_class                          | test_method                                  | test_arguments                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                             |
|:------------------------------------|:---------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests       | test_replication_basic                       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74086#0199de9d-aca9-4078-a63f-b338efc8c20f | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic       |
| ShadowLinkingReplicationTests       | test_replication_basic                       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74086#0199dea4-6454-4058-8be0-3888cf7798a4 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic       |
| ShadowLinkingReplicationTests       | test_replication_basic                       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/74086#0199de9d-acaa-4019-a411-1a6a06d7ef1d | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic       |
| ShadowLinkingReplicationTests       | test_replication_basic                       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/74086#0199dea4-6456-49a2-9ac7-ccd18ec0d7f7 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic       |
| SegmentMsTest                       | test_segment_rolling_with_retention_consumer | null                                                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/74086#0199de9d-acae-407c-b60d-c924d6af1362 | FLAKY         | 18/21    | upstream reliability is '91.82879377431907'. current run reliability is '85.71428571428571'. drift is 6.11451 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| WriteCachingFailureInjectionE2ETest | test_crash_all                               | {"use_transactions": false}                                                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/74086#0199dea4-645b-4f42-848c-f4b5058aa874 | FLAKY         | 18/21    | upstream reliability is '82.708142726441'. current run reliability is '85.71428571428571'. drift is -3.00614 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all         |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74101">build#74101</a></summary>


| test_class                      | test_method                  | test_arguments                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                           |
|:--------------------------------|:-----------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests   | test_replication_basic       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74101#0199df8f-63fe-4164-9f0e-8b5168385c86 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic     |
| ShadowLinkingReplicationTests   | test_replication_basic       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74101#0199df97-7d09-4d85-8d05-ad1a5e154aff | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic     |
| ShadowLinkingReplicationTests   | test_replication_basic       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/74101#0199df8f-63ff-49ed-8580-ec276859db61 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic     |
| ShadowLinkingReplicationTests   | test_replication_basic       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/74101#0199df97-7d0c-4792-9594-89ad99456809 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic     |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test     | {"cloud_storage_type": 1, "mixed_versions": true}                                                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/74101#0199df8f-63ff-49ed-8580-ec276859db61 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| TimeQueryTest                   | test_timequery_with_local_gc | null                                                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/74101#0199df8f-63fe-4164-9f0e-8b5168385c86 | FLAKY         | 18/21    | upstream reliability is '89.86371379897786'. current run reliability is '85.71428571428571'. drift is 4.14943 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery_with_local_gc               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74227">build#74227</a></summary>


| test_class                    | test_method                     | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                       |
|:------------------------------|:--------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests          | test_rapid_shadow_link_toggling | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74227#0199e557-595e-422d-a6cf-21e36a18c50a | FLAKY         | 19/21    | upstream reliability is '98.56115107913669'. current run reliability is '90.47619047619048'. drift is 8.08496 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling |
| ShadowLinkingReplicationTests | test_replication_basic          | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74227#0199e557-5963-4218-a99e-ab9a60435517 | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic |
| QuickTerminateTest            | test_terminate                  | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74227#0199e557-5966-4b09-9adf-0a3736e9a6cb | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=QuickTerminateTest&test_method=test_terminate                    |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74354">build#74354</a></summary>


| test_class                    | test_method                                             | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                                   |
|:------------------------------|:--------------------------------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_replication_basic                                  | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74354#0199ef91-595e-423f-9a5d-205f6fb63172 | FLAKY         | 15/21    | upstream reliability is '100.0'. current run reliability is '71.42857142857143'. drift is 28.57143 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic                             |
| ShadowLinkingReplicationTests | test_replication_with_failures                          | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74354#0199ef95-46d0-4a79-9e2d-d21bd465d1fd | FLAKY         | 11/21    | upstream reliability is '100.0'. current run reliability is '52.38095238095239'. drift is 47.61905 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures                     |
| NodesDecommissioningTest      | test_decommissioning_crashed_node                       | {"cloud_topic": true}                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/74354#0199ef91-595c-4326-b3c4-41ce1cb2f0c3 | FLAKY         | 11/21    | upstream reliability is '84.31372549019608'. current run reliability is '52.38095238095239'. drift is 31.93277 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_crashed_node                       |
| NodesDecommissioningTest      | test_decommissioning_crashed_node                       | {"cloud_topic": true}                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/74354#0199ef95-46ce-4b5d-a767-84b1e86bec25 | FLAKY         | 13/21    | upstream reliability is '84.31372549019608'. current run reliability is '61.904761904761905'. drift is 22.40896 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_crashed_node                       |
| NodesDecommissioningTest      | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": true, "delete_topic": false}                                      | integration | https://buildkite.com/redpanda/redpanda/builds/74354#0199ef91-595e-423f-9a5d-205f6fb63172 | FLAKY         | 9/21     | upstream reliability is '90.12345679012346'. current run reliability is '42.857142857142854'. drift is 47.26631 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| NodesDecommissioningTest      | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": true, "delete_topic": false}                                      | integration | https://buildkite.com/redpanda/redpanda/builds/74354#0199ef95-46d0-4c39-80f4-eb8b4515eaa4 | FLAKY         | 15/21    | upstream reliability is '83.33333333333334'. current run reliability is '71.42857142857143'. drift is 11.90476 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| NodesDecommissioningTest      | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": true, "delete_topic": true}                                       | integration | https://buildkite.com/redpanda/redpanda/builds/74354#0199ef95-46d1-4122-9cd9-3b57bd2218ae | FLAKY         | 12/21    | upstream reliability is '89.92805755395683'. current run reliability is '57.14285714285714'. drift is 32.7852 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| NodesDecommissioningTest      | test_decommissioning_working_node                       | {"cloud_topic": true, "delete_topic": false, "tick_interval": 3600000}            | integration | https://buildkite.com/redpanda/redpanda/builds/74354#0199ef91-595c-4d52-9758-3cc28cd29e62 | FLAKY         | 15/21    | upstream reliability is '83.16831683168317'. current run reliability is '71.42857142857143'. drift is 11.73975 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_working_node                       |
| NodesDecommissioningTest      | test_decommissioning_working_node                       | {"cloud_topic": true, "delete_topic": false, "tick_interval": 5000}               | integration | https://buildkite.com/redpanda/redpanda/builds/74354#0199ef95-46cf-406f-8238-aed69bb5aa72 | FLAKY         | 12/21    | upstream reliability is '82.35294117647058'. current run reliability is '57.14285714285714'. drift is 25.21008 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_working_node                       |
| NodesDecommissioningTest      | test_decommissioning_working_node                       | {"cloud_topic": true, "delete_topic": true, "tick_interval": 3600000}             | integration | https://buildkite.com/redpanda/redpanda/builds/74354#0199ef91-595f-4056-996c-51450c5dc382 | FLAKY         | 7/21     | upstream reliability is '83.60655737704919'. current run reliability is '33.33333333333333'. drift is 50.27322 and the allowed drift is set to 50. The test should FAIL  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_working_node                       |
| NodesDecommissioningTest      | test_recommissioning_node                               | {"cloud_topic": true}                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/74354#0199ef95-46ce-4b5d-a767-84b1e86bec25 | FLAKY         | 14/21    | upstream reliability is '80.32786885245902'. current run reliability is '66.66666666666666'. drift is 13.6612 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_node                               |
| NodesDecommissioningTest      | test_recommissioning_node_finishes                      | {"cloud_topic": true}                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/74354#0199ef95-46d0-4c39-80f4-eb8b4515eaa4 | FLAKY         | 11/21    | upstream reliability is '100.0'. current run reliability is '52.38095238095239'. drift is 47.61905 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_node_finishes                      |
| NodesDecommissioningTest      | test_recommissioning_one_of_decommissioned_nodes        | {"cloud_topic": true}                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/74354#0199ef91-595f-4f2d-bb20-9a3fb81b2f2b | FLAKY         | 14/21    | upstream reliability is '85.29411764705883'. current run reliability is '66.66666666666666'. drift is 18.62745 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_one_of_decommissioned_nodes        |
</details>
<!--/body-->

michael-redpanda commented:
Force push:

* Rebased off of dev

michael-redpanda commented:
Force push:

* Rebased off of dev

michael-redpanda commented:
Force push:

* Another rebase

bharathv commented on src/v/cluster_link/replication/types.cc:
```diff
@@ -0,0 +1,22 @@
+      "{{source_hwm: {}, source_lso: {}, update_time: {}}}",
+      source_hwm,
+      source_lso,
+      update_time.time_since_epoch().count());
```
nit: print in ms or some human parseable units.

bharathv commented on src/v/cluster_link/replication/partition_data_queue.h:
```diff
@@ -55,10 +61,18 @@ class partition_data_queue {
     void do_reset(kafka::offset next);
     kafka::offset _next{};
+    kafka::offset _source_hwm{-1};
+    kafka::offset _source_lso{-1};
```
nit: last_known_source_hwm/ last_known_source_lso ?

michael-redpanda replied:
`source_hwm` is the last known source hwm right?

michael-redpanda replied:
n/m renaming nit got it

bharathv commented on src/v/cluster_link/service.cc:
```diff
@@ -729,6 +729,7 @@ service::shadow_topic_report(model::id_t link_id, const ::model::topic& topic) {
                           "Error getting shadow topic report from node {}: {}",
                           node_id,
                           r.err_code);
+                        return ss::now();
```
wow.

bharathv commented on src/v/cluster_link/replication/partition_data_queue.h:
```diff
@@ -34,6 +34,12 @@ namespace cluster_link::replication {
+        kafka::offset source_hwm;
+        kafka::offset source_lso;
+        ss::lowres_clock::time_point fetch_time;
+    };
```
was wondering if we track these stats in fetch_data {} struct and pass it along to the sink and track everything there .. sink is the one stop shop to all metrics including hwm (especially since partition_data_queue is an internal abstraction and would be nice to keep it shielded from all these details.)

michael-redpanda replied:
i'll look into it

bharathv commented on src/v/cluster_link/replication/link_replication_mgr.cc:
```diff
@@ -137,4 +137,26 @@ void link_replication_manager::stop_replicators(
+    }
+    return it->second->get_source_partition_offsets();
+}
 } // namespace cluster_link::replication
```
think this is missing shadow partition hwm.. also can we call them partition_stats{} or something (source ..offsets seems too specific).

michael-redpanda replied:
will do

bharathv commented on src/v/cluster_link/manager.h:
```diff
@@ -181,6 +181,14 @@ class manager {
     std::unique_ptr<link_registry>& registry() noexcept { return _registry; }
 
+    cl_result<
+      chunked_hash_map<::model::ntp, replication::source_partition_offsets>>
```
nit s/ shadow_partition_stats or some such.

michael-redpanda replied:
`partition_offsets_report`?

bharathv commented on src/v/cluster_link/model/types.h:
```diff
@@ -978,6 +978,9 @@ struct shadow_topic_partition_leader_report
     // checkpointing
+    kafka::offset source_partition_high_watermark;
+    kafka::offset source_partition_last_stable_offset;
+    std::int64_t last_update_time_ms_since_epoch;
```
chrono::milliseconds?

michael-redpanda replied:
yeah that would be the smarter thing to do lol

bharathv commented on src/v/cluster_link/service.cc:
```diff
@@ -858,4 +860,72 @@ service::shard_local_shadow_link_report(model::id_t id) {
+    if (!link_id.has_value()) {
+        co_return std::unexpected<errc>(errc::link_id_not_found);
+    }
+    auto& members_table = _controller->get_members_table();
+    const auto& node_ids = members_table.local().node_ids();
+    vlog(cllog.trace, "Issuing rpcs to nodes {}", node_ids);
+    model::shadow_link_status_report results;
```
IIUC, I think the way this (reducer) works even if one of the nodes is down, the report generation hits an error. Couple of things I can think of.

- the list of nodes should be a union of all the nodes that host replicas of topics belonging to the link.
- Even if there is a partial error in the link report collection, I think the partial report should be bubbled back to the user (which enough context that the collection ran into an errror and hence the report is partial).

michael-redpanda replied:
ok sg

michael-redpanda replied:
So if the node error's out, we log the error but doesn't have an effect on the final report.

michael-redpanda replied:
And for the per-shard report, the error that would override everything would be if the link couldn't be found.

bharathv replied:
> we log the error but doesn't have an effect on the final report.

should the overall report ec should still be set? (indicating it is partial or sth like that).

oleiman commented on src/v/redpanda/admin/services/shadow_link/converter.cc:
```diff
@@ -832,23 +833,10 @@ create_shadow_topics(const cluster_link::model::link_state& state) {
+    std::ranges::transform(
+      state.mirror_topics,
+      std::back_inserter(shadow_topics),
+      [](const auto& p) { return model_to_shadow_topic(p.first, p.second); });
```
❤️

michael-redpanda committed:
cl: Added method to return sink HWM

Added method to data_sink to return the associated partition's HWM.
This will be returned in the offset report.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Added method to get source partition offsets

Added method to the data_source interface to get the source partitions
offsets.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Add methods to get partition offset report

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/client: Return log_start_offset

Return the partition's log_start_offset in the fetch response data.  To
be used in partition reports

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Addressed PR comments
* Now returning shadow partition's HWM
* Tracking source partition's log start offset

michael-redpanda committed:
k/c/dc: Add last_known_source_offsets to subscription data

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/c/dc: Add method to update source partitions

This method will be used by fetcher to update the source partitions'
last known offsets

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/c/dc: Update source partitions on DC from fetcher

When fetcher handles a successful fetch response, update the source
partition offsets held by the direct_consumer.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/c/dc: Add methods to get source offsets

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/replication: Add method to get source offsets

Add ability to get last recorded source offsets from the underlying
direct_consumer.  Used by data_source

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Addressed issues with getting offsets from fetcher out

michael-redpanda commented:
Force push:

* Rebased off of dev

michael-redpanda commented:
Force push:

* Fixed cpp lint

bharathv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -541,17 +541,27 @@ fetcher::process_fetch_response(
                 part_data.aborted_transactions = std::move(
                   part_response.aborted_transactions);
 
+                _parent->maybe_update_source_partition_offsets(
```
I'm wondering if we should pass the subscription epoch  to this function (to ensure the epoch is still valid at the time of updating).

cc: @joe-redpanda can you please review these offset changes, thanks. The idea is to the track the partition low/high wm and lso as data flows through the fetchers. (they are used by by the status APIs to populate the link topic statuses in later commits).

joe-redpanda replied:
I think even with comparison we have a race

initial state t/p0 offsets [100,200) fetcher epoch 0, subscription epoch 0

100 batches are given to the source cluster, once accepted, the operator truncates (-inf, 197]

fetch -> new offsets [200,300), prefix truncate 197 & below

fetch is consistent, gets accepted which
1. places [200,300) on the queue
2. indicates truncation to direct_consumer which now shows [198,200)

truncation task picks this up and truncates

Someone has an oopsie, the fetch offset needs to get reset back to 200, subscription epoch 0->1

our [200,300) is now next up to be pulled from the queue and gets filtered out

source cluster dies, the next fetches are never coming

we are now in an odd state that never existed on the source cluster which is [198,200)

joe-redpanda replied:
IMO from the DC side, the best thing to do would be to attach this to the nonsense that goes in the data_queue s.t. the decision on whether to accept or discard a fetch is transactional. Either we accept the truncation AND batches, or we throw both away.

bharathv replied:
> we are now in an odd state that never existed on the source cluster which is [198,200)

Isn't this acceptable since the source cluster is dead? Once it comes up we the target will catch up to [198 - 300] or whatever the correct hwm is?

That aside, the offsets being populated are always consistent/source of truth, right? (because they are a part of fetch response)

joe-redpanda replied:
> Isn't this acceptable since the source cluster is dead? 

Not sure. I think thats more a product question.

> That aside, the offsets being populated are always consistent/source of truth, right? 

They're consistent / true at some point in time. I think what bothers me here is that we are breaking the consistent truth of a fetch into pieces, and ingesting them piecewise/non-atomically. Our target cluster can then be a frankenstein of different point-in-time truths.

The swap to putting this on the queue shouldn't be too too much additional work in exchange for what I think is a lot less possible edge cases.

bharathv replied:
We initially considered adding this to the queue, but that would require further changes I think. The queue has a size limit, and a large burst of metadata updates could potentially stall it.. although that's unlikely. One option could be to treat these metadata entries as zero-length (similar to how errors are treated as zero size), but that would involve some refactoring. IMO, since this API is currently used internally and doesn't expect a specific processing order (i.e., metadata isn’t required to be consumed after data replication), the current approach is ok. We could explore a follow-up change to enqueue metadata updates in a way that doesn't affect queue length once the status API/metrics land. (Lets see what the team thinks)

(chatted with Joe offline).

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -108,6 +108,25 @@ void direct_consumer::update_configuration(configuration cfg) {
+direct_consumer::get_source_offsets(const model::topic_partition& tp) const {
+    return get_source_offsets(
+      model::topic_partition_view(tp.topic, tp.partition));
+}
```
nit: the caller can do this? (do we need both)

oleiman replied:
Incidentally, idk why we don't have `topic_partition_view::topic_partition_view(const topic_partition&)`

bharathv commented on proto/redpanda/core/admin/v2/shadow_link.proto:
```diff
@@ -421,6 +437,8 @@ message TopicPartitionInformation {
+    // Timestamp of the last time the source partition information was updated
+    google.protobuf.Timestamp source_last_updated_timestamp = 5;
     // Shadowed partition's HWM
     int64 high_watermark = 4;
```
nit: flip this (for readability)

bharathv commented on src/v/cluster_link/replication/partition_replicator.cc:
```diff
@@ -163,6 +163,25 @@ ss::future<> partition_replicator::replicate_and_wait(
+      .source_lso = source_info.has_value() ? source_info->source_lso
+                                            : kafka::offset{},
+      .update_time = source_info.has_value() ? source_info->update_time
+                                             : ss::lowres_clock::time_point{},
```
can we use a non existent value? Using now() for missing info seems a bit off.

michael-redpanda replied:
It's not `now`?  It uses default which is 0

bharathv replied:
duh, sorry read it as now().

bharathv commented on src/v/cluster_link/model/types.cc:
```diff
@@ -286,7 +286,18 @@ fmt::iterator shadow_topic_report_request::format_to(fmt::iterator it) const {
+      source_partition_start_offset,
+      source_partition_high_watermark,
+      source_partition_last_stable_offset,
+      last_update_time,
```
I think now() - time in ms is probably more human readable.

oleiman replied:
could render epoch time?

michael-redpanda replied:
Sorry not entirely following the suggestion here.  Are you looking for a nicely formatted time point?

oleiman replied:
eh, ignore me. not 100% sure that's what Bharath was driving at.

bharathv commented on src/v/cluster_link/model/types.cc:
```diff
@@ -348,6 +348,21 @@ shadow_link_status_report_response::format_to(fmt::iterator it) const {
+        first = false;
+        fmt::format_to(it, "{}: {}", topic, response);
+    }
+    return fmt::format_to(it, "] }}");
```
I think this will cross allocation thresholds at scale.

michael-redpanda replied:
yeah that's fair... do we have any prior art I can reference?

bharathv commented on src/v/cluster_link/service.cc:
```diff
@@ -74,6 +74,41 @@ struct shard_report_reducer {
+        if (result->err_code != ::cluster_link::errc::success) {
+            // once we have an error, just keep it
+            return;
+        }
```
I wonder if we should do this.. (if the shard_result is successful, we can still update the reports with the results from the shard, even tho the whole report is partial).

bharathv commented on tests/rptest/tests/cluster_linking_e2e_test.py:
```diff
@@ -1031,6 +1034,61 @@ def test_replication_basic(self, shuffle_leadership, source_cluster_spec):
+        ) -> tuple[bool, shadow_link_pb2.ShadowTopic | None]:
+            shadow_topic = self.get_shadow_topic(
+                shadow_link_name="test-link", shadow_topic_name=topic.name
+            )
```
can we wire the same checks into "test_replication_with_failures".. I think that exercises more interesting paths of report collection code.

michael-redpanda commented:
Force push:

* Address clang-tidy bug

michael-redpanda commented:
Force push:

* Rebased off of dev

oleiman commented on src/v/kafka/client/direct_consumer/api_types.h:
```diff
@@ -80,6 +80,13 @@ struct fetched_topic_data {
 
 using fetches = kafka_result<chunked_vector<fetched_topic_data>>;
 
+struct source_partition_offsets {
+    kafka::offset log_start_offset{-1};
+    kafka::offset high_watermark{-1};
+    kafka::offset last_stable_offset{-1};
+    ss::lowres_clock::time_point last_offset_update_timestamp{};
+};
```
a block comment would be helpful here for context.

oleiman commented on src/v/cluster_link/replication/deps.h:
```diff
@@ -52,6 +52,22 @@ class data_sink_factory {
+              source_hwm,
+              source_lso,
+              update_time.time_since_epoch().count());
+        }
```
nitpick: does this need to be in the header?

michael-redpanda replied:
no - there wasn't a `deps.cc`, I could just make one lol

oleiman commented on src/v/cluster_link/replication/link_replication_mgr.cc:
```diff
@@ -137,4 +137,23 @@ void link_replication_manager::stop_replicators(
 }
 
+chunked_hash_map<::model::ntp, partition_offsets_report>
+link_replication_manager::get_partition_offsets_report() const {
+    chunked_hash_map<::model::ntp, partition_offsets_report> results;
+
```
nitpick: `results.reserve(_replicators.size())`

oleiman commented on src/v/cluster_link/model/types.h:
```diff
@@ -1012,6 +1012,11 @@ struct shadow_topic_partition_leader_report
     // todo: add offset information for promotion
     // todo: add hwm information for fail over state
```
are these satisfied now?

oleiman commented on src/v/cluster_link/model/types.cc:
```diff
@@ -309,6 +309,43 @@ fmt::iterator shadow_topic_report_response::format_to(fmt::iterator it) const {
+        first = false;
+        fmt::format_to(it, "{}: {}", pid, report);
+    }
+    return fmt::format_to(it, "] }}");
```
```suggestion
    fmt::format_to(
        it, "{{ status: {}, partition_reports: {}, }}",
        status, 
        fmt::join(partition_reports | std::views::transform([](const auto r) {
            return fmt::format("{}: {}", r.pid, r.report);
        }, ", "));
```

ish?

oleiman commented on src/v/redpanda/admin/services/shadow_link/shadow_link.cc:
```diff
@@ -39,27 +34,49 @@ T handle_error(cluster_link::cl_result<T> result) {
+    auto code = info.code();
+    handle_error(code, std::move(info).message());
+    // Handle error should throw
+    __builtin_unreachable();
```
nitpick

```suggestion
    std::unreachable;
```

oleiman reviewed (commented):
bit late to the party, but looks pretty much legit to me modulo a few nits.

michael-redpanda commented:
> bit late to the party, but looks pretty much legit to me modulo a few nits.

@oleiman you're always welcome!  hope you brought snacks

michael-redpanda commented:
Force push:

* Addressed some PR comments

rockwotj commented on proto/redpanda/core/admin/v2/shadow_link.proto:
```diff
@@ -29,6 +29,7 @@ import "google/protobuf/field_mask.proto";
 option (redpanda.core.pbgen.cpp_namespace) = "proto::admin";
 
 service ShadowLinkService {
+    // RPCs for the ShadowLink resource
```
This should be a comment for `CreateShadowLink` otherwise the docs will render this as a comment on the `CreateShadowLink` API. Let's just remove it.

rockwotj commented on proto/redpanda/core/admin/v2/shadow_link.proto:
```diff
@@ -64,6 +65,20 @@ service ShadowLinkService {
         };
     }
+
+    // RPCs for the ShadowTopic resource
+
+    rpc GetShadowTopic(GetShadowTopicRequest) returns (GetShadowTopicResponse) {
+        option (redpanda.core.pbgen.rpc) = {
+            authz: SUPERUSER
+        };
+    }
+    rpc ListShadowTopics(ListShadowTopicsRequest)
+        returns (ListShadowTopicsResponse) {
+        option (redpanda.core.pbgen.rpc) = {
+            authz: SUPERUSER
+        };
+    }
```
I know this is just CRUD, but since it's a read only resource comments about what you are getting here, how the are created and why you would want to call these APIs is useful

michael-redpanda commented:
Force push:

* Improved commenting on protos

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74354-->
<!--e501ebce-8732-437e-a823-9c64c17c9f04-->
<!--title-->
## Retry command for [Build#74354](https://buildkite.com/redpanda/redpanda/builds/74354)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_decommissioning_working_node@{"cloud_topic":true,"delete_topic":true,"tick_interval":3600000}
```
<!--/body-->

michael-redpanda commented:
CI Failure:
```
Module:      rptest.tests.nodes_decommissioning_test
Class:       NodesDecommissioningTest
Method:      test_decommissioning_working_node
Nodes (used/allocated): 6/6
Arguments:
{
  "cloud_topic": true,
  "delete_topic": true,
  "tick_interval": 3600000
}
```

known flake

dotnwat commented on src/v/cluster_link/service.cc:
```diff
@@ -752,4 +785,150 @@ service::shadow_topic_report(model::id_t link_id, const ::model::topic& topic) {
+    model::shadow_link_status_report results;
+    results.link_id = link_id.value();
+    try {
+        co_await ss::max_concurrent_for_each(
+          node_ids,
```
doesn't this have a small chance of hitting a concurrent modification to the node_ids set? presumably it's very small (number of brokers) and we could just make a copy to operate on here?
